### PR TITLE
[docs] update CPU dockerfiles (fixes #5550)

### LINKED ISSUE #5550: cmake error at linking process while integrating lightgbm
## Description

I want to integrate lightgbm into my project to support local lightgbm model inference, but I meet an error at the process 
 of linking my executable target to the static lib(lib_lightgbm.a) compiled by cmake on lightgbm.

Here is the .cmake file to integrate lightgbm:

lightgbm.cmake
```cmake
INCLUDE(ExternalProject)

SET(LIGHTGBM_SOURCES_DIR ${THIRD_PARTY_PATH}/lightgbm)
SET(LIGHTGBM_INSTALL_DIR ${THIRD_PARTY_PATH}/install/lightgbm)
SET(LIGHTGBM_INCLUDE_DIR "${LIGHTGBM_INSTALL_DIR}/include" CACHE PATH "lightgbm include directory." FORCE)
SET(LIGHTGBM_LIBRARIES "${LIGHTGBM_INSTALL_DIR}/lib/lib_lightgbm.a" CACHE FILEPATH "LIGHTGBM_LIBRARIES" FORCE)

INCLUDE_DIRECTORIES(${LIGHTGBM_INCLUDE_DIR})

if (EXISTS ${LIGHTGBM_LIBRARIES})
    MESSAGE(STATUS "LIGHTGBM_LIBRARIES: ${LIGHTGBM_LIBRARIES}")
else ()
    ExternalProject_Add(
            extern_lightgbm # -llet
            GIT_REPOSITORY https://github.com/microsoft/LightGBM.git
            GIT_TAG a0a617dfff2e2e1637b72b7a2adbe5453e7f97e0
            PREFIX ${LIGHTGBM_SOURCES_DIR}
            CMAKE_ARGS
            -DCMAKE_BUILD_TYPE=Release
            -DCMAKE_INSTALL_PREFIX=${LIGHTGBM_INSTALL_DIR} #
            -DDISABLE_TESTS=ON
            #-DBUILD_SHARED_LIBS=ON
            -DBUILD_STATIC_LIB=ON
    )
endif ()

ADD_LIBRARY(lightgbm STATIC IMPORTED GLOBAL)
SET_PROPERTY(TARGET lightgbm PROPERTY IMPORTED_LOCATION ${LIGHTGBM_LIBRARIES})
ADD_DEPENDENCIES(lightgbm extern_lightgbm)

LIST(APPEND external_project_dependencies lightgbm)
```
The cmake build dir structure generated is as follows. The external_libs dir (as I marked, which is the git submodule in lightgbm project) is copyed to there manually, or there will ocurr an error of missing dependent .h file in external_libs when I import the header files in lightgbm include dir (in the cmake install dir).
image

full cmake install dir:

image

main.cpp
```cpp
#include <LightGBM/utils/common.h>
#include <LightGBM/c_api.h>


int main(int argc, char **argv) {
  BoosterHandle out;
  int out_num_iterations;
  const char* model_file_name = "/root/model_0919.txt";
  int code = LGBM_BoosterCreateFromModelfile(model_file_name, &out_num_iterations, nullptr);
}
```

## Reproducible example

Below is the cmake error output at the last linking process. As the output is too long, I cut out the last few lines. It appears that some function implement was not found in lib_lightgbm.a lib.
```shell
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o):data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x103): more undefined references to `GOMP_parallel' follow
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::CUDATreeLearner>::FindBestSplits(LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x12): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x19): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::GPUTreeLearner>::FindBestSplits(LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x12): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x19): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::GPUTreeLearner>::FindBestSplits(LightGBM::Tree const*) [clone ._omp_fn.1]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.1[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x10): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.1[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x17): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::SerialTreeLearner>::FindBestSplits(LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x12): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x19): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::SerialTreeLearner>::FindBestSplits(LightGBM::Tree const*) [clone ._omp_fn.1]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.1[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x10): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE._omp_fn.1[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE14FindBestSplitsEPKNS_4TreeE]+0x17): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::CUDATreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x1c): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x24): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::GPUTreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x1c): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x24): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::SerialTreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*) [clone ._omp_fn.0]':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x1c): undefined reference to `omp_get_num_threads'
/usr/bin/ld: data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE._omp_fn.0[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x24): undefined reference to `omp_get_thread_num'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::CUDATreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*)':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE[_ZN8LightGBM23DataParallelTreeLearnerINS_15CUDATreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x622): undefined reference to `GOMP_parallel'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::GPUTreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*)':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE[_ZN8LightGBM23DataParallelTreeLearnerINS_14GPUTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x622): undefined reference to `GOMP_parallel'
/usr/bin/ld: ../../third_party/install/lightgbm/lib/lib_lightgbm.a(data_parallel_tree_learner.cpp.o): in function `LightGBM::DataParallelTreeLearner<LightGBM::SerialTreeLearner>::FindBestSplitsFromHistograms(std::vector<signed char, std::allocator<signed char> > const&, bool, LightGBM::Tree const*)':
data_parallel_tree_learner.cpp:(.text._ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE[_ZN8LightGBM23DataParallelTreeLearnerINS_17SerialTreeLearnerEE28FindBestSplitsFromHistogramsERKSt6vectorIaSaIaEEbPKNS_4TreeE]+0x622): undefined reference to `GOMP_parallel'
collect2: error: ld returned 1 exit status
make[3]: *** [src/rankserver/CMakeFiles/rank_server_exec.dir/build.make:150: src/rankserver/rank_server_exec] Error 1
make[2]: *** [CMakeFiles/Makefile2:4492: src/rankserver/CMakeFiles/rank_server_exec.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:4499: src/rankserver/CMakeFiles/rank_server_exec.dir/rule] Error 2
```

## Environment info

LightGBM version or commit hash: v3.3.3

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #5550.

The Dockerfiles in `docker/` can be used to build container images for running different builds of LightGBM. It's 

This PR proposes some changes to the non-GPU Dockerfiles there, to hopefully make them easier to use, especially for people who are new to LightGBM.

* updates from Ubuntu 16.04 ([entered only-getting-security-patches mode in 2021](https://ubuntu.com/about/release-cycle))  to Ubuntu 20.04
* installs OpenMP in every image
* ensures that you can copy, paste, and run examples with 0 modification
* switches from CMake-based build of the R package to CRAN-style package, since it's less likely to have installation issues
* other misc. minor cleanup (like indentation so flags in `docker run` are easier to read)

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -12,113 +14,183 @@ Follow the general installation instructions [on the Docker site](https://docs.d
...
-task=convert_model
-input_model=model.txt
-convert_model=/out/predict.cpp
-convert_model_language=cpp
+# train, and save model to a text file
+docker run \
+  --rm \
+  --volume "${PWD}":/opt/training \
+  --workdir /opt/training \
+  -it lightgbm-cli \
```
Comment (jmoralez) [inline docker/README.md:50]:
```suggestion
  lightgbm-cli \
```

File: docker/README.md [hunk 2]
```diff
@@ -12,113 +14,183 @@ Follow the general installation instructions [on the Docker site](https://docs.d
...
+)
+bst.save_model("LightGBM-python-model.txt")
+EOF
+
+# run training in a container
+docker run \
+    --rm \
+    --volume "${PWD}":/opt/training \
+    --workdir /opt/training \
+    -it lightgbm-python \
```
Comment (jmoralez) [inline docker/README.md:100]:
```suggestion
    lightgbm-python \
```

File: docker/README.md [hunk 3]
```diff
@@ -12,113 +14,183 @@ Follow the general installation instructions [on the Docker site](https://docs.d
...
+)
+lgb.save(bst, "LightGBM-R-model.txt")
+EOF
+
+# run training in a container
+docker run \
+    --rm \
+    --volume "${PWD}":/opt/training \
+    --workdir /opt/training \
+    -it lightgbm-r \
```
Comment (jmoralez) [inline docker/README.md:159]:
```suggestion
    lightgbm-r \
```

File: docker/README.md [hunk 4]
```diff
@@ -12,113 +14,183 @@ Follow the general installation instructions [on the Docker site](https://docs.d
...
-docker run --rm -it -e PASSWORD=lightgbm -p 8787:8787 lightgbm-r
+
+To use [RStudio](https://www.rstudio.com/products/rstudio/), an interactive development environment, run the following.
+
+```shell
+docker run \
+    --rm \
+    --env PASSWORD="lightgbm" \
+    -p 8787:8787 \
+    -it lightgbm-r
```
Comment (jmoralez) [inline docker/README.md:181]:
```suggestion
    lightgbm-r
```

**Review by jmoralez — APPROVED** [review]
Thanks! LGTM. Just some optional removals of the interactive terminal.

**Comment by jameslamb** [comment]
> removals of the interactive terminal

Thanks for that! You're right, for these cases the `-it` for `docker run` was unnecessary. Just applied those suggestions, I'll merge this once CI builds.

**Pull request merged**